WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCouponViewModel.swift

```diff
@@ -164,6 +169,13 @@ final class AddEditCouponViewModel: ObservableObject {
     }
 
     func updateCoupon(coupon: Coupon) {
+        if let validationError = validateCouponLocally(coupon) {
```

Comment:
I would love to know your thoughts on making `func validateCouponLocally(_ coupon: Coupon)` throw an error instead of returning one?

Something like the following,
```swift
do {
    try validateCouponLocally(coupon)
} catch {
    if let error = error as? CouponError {
        notice = NoticeFactory.createCouponErrorNotice(error,
                                                       editingOption: editingOption)
    }
    onCompletion?(.failure(error))
}
```

```swift
func validateCouponLocally(_ coupon: Coupon) throws {
    if coupon.code.isEmpty {
        throw CouponError.couponCodeEmpty
    }
}
````

Reply:
I found it more "expensive" in terms of lines of code and readability, and you need to always add a `try-catch` also where maybe you don't need it (like in tests). What are the benefits to use this approach in your opinion?

Reply:
Thanks for sharing your thoughts. I totally agree with the `lines of code` point. 👍 

IMHO, 

**Good**
- Making this function throw an error instead of returning an error makes the code clear and readable. 
- We can even return a `Bool` and throw an error. i.e. `func isCouponValid(_ coupon: Coupon) throws -> Bool`

**Bad**
- We have to cast the error from `Error` into `CouponError`. With the current approach, the compiler takes care of that. 😄 

-------------

> and you need to always add a try-catch also where maybe you don't need it (like in tests

I agree. 👍 Meanwhile, I think we can re-write the tests into, 

- A test that throws.
```swift
func test_validateCouponLocally_throws_expected_error_if_coupon_code_is_empty() throws {
    // Given
    let coupon = Coupon.sampleCoupon.copy(code: "")
    let viewModel = AddEditCouponViewModel(existingCoupon: coupon)

    // When
    do {
        try viewModel.validateCouponLocally(coupon)
    } catch {
        guard let error = error as? AddEditCouponViewModel.CouponError else {
            XCTFail("validateCouponLocally is throwing unexpected error.")
            return
        }

        // Then
        XCTAssertEqual(error, .couponCodeEmpty)
    }
}
```
- A test that uses `XCTAssertNoThrow`.
```swift
func test_validateCouponLocally_does_not_throw_an_error_if_coupon_code_is_not_empty() {
    // Given
    let coupon = Coupon.sampleCoupon.copy(code: "ABCDEF")
    let viewModel = AddEditCouponViewModel(existingCoupon: coupon)

    // When & Then
    XCTAssertNoThrow(try viewModel.validateCouponLocally(coupon))
}
```

-----

I am totally fine with keeping the `returning error` approach. 👍😄